docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -157,6 +157,94 @@ Pros:
 Cons:
 1. Cannot wrap antehandlers with decorators like you can with Weave.
 
+### Simple Decorators
+
+This approach takes inspiration from Weave's decorator design while trying to minimize the number of breaking changes to the SDK and maximizing simplicity.
+
+Allow Decorator interface that can be chained together to create an SDK AnteHandler.
+
+This allows users to choose between implementing an AnteHandler by themselves and setting it in the baseapp, or use the decorator pattern to chain their custom decorators with SDK provided decorators in the order they wish.
+
+```golang
+// A Decorator wraps an AnteHandler, and can do pre- and post-processing on the next AnteHandler
+type Decorator interface {
+    AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error)
+}
+```
+
+```golang
+// ChainDecorators will recursively link all of the Decorators in the chain and return a final AnteHandler function
+// This is done to preserve the ability to set a single AnteHandler function in the baseapp.
+func ChainDecorators(chain ...Decorator) AnteHandler {
+    if len(chain) == 1 {
+        return func(ctx Context, tx Tx, simulate bool) {
+            chain[0].AnteHandle(ctx, tx, simulate, nil)
+        }
+    }
+    return func(ctx Context, tx Tx, simulate bool) {
+        chain[0].AnteHandle(ctx, tx, simulate, ChainDecorators(chain[1:]))
+    }
+}
+```
+
+#### Example Code
+
+```golang
+// Setup GasMeter, catch OutOfGasPanic and handle appropriately
+type SetupDecorator struct{}
+
+func (sud SetupDecorator) AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error) {
+    ctx.GasMeter = NewGasMeter(tx.Gas)
+
+    defer func() {
+        // recover from OutOfGas panic and handle appropriately
+    }
+
+    return next(ctx, tx, simulate)
+}
+
+// Signature Verification decorator. Verify Signatures and move on
+type SigVerifyDecorator struct{}
+
+func (svd SigVerifyDecorator) AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error) {
+    // verify sigs. Return error if invalid
+
+    // call next antehandler if sigs ok
+    return next(ctx, tx, simulate)
+}
+
+// User-defined Decorator. Can choose to pre- and post-process on AnteHandler
+type UserDefinedDecorator struct{
+    // custom fields
+}
+
+func (udd UserDefinedDecorator) AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error) {
+    // pre-processing logic
+
+    ctx, err = next(ctx, tx, simulate)
+
+    // post-processing logic
+}
+```
+
+```golang
+// Create final antehandler by chaining the decorators together
+antehandler := ChainDecorators(NewSetupDecorator(), NewSigVerifyDecorator(), NewUserDefinedDecorator())
+
+// Set chained Antehandler in the baseapp
+bapp.SetAnteHandler(antehandler)
+```
+
+Pros:
+
+1. Allows one decorator to pre- and post-process the next AnteHandler, similar to the Weave design.
```

Comment:
There is no concept of `AnteHandler(); Handler()` in weave.
It is `Decorator(Handler)`.

It is not just the AnteHandler. The enormous body of DeliverTx, runTx and runMsg in BaseApp show all the code that cannot be customized. *Exactly* because there is no other place to put code that runs both before and after all handlers.

Reply:
I agree all of the functions you mentioned above could do with a refactor that makes them simpler and more readable. I'm not convinced that anything they do should be customizable though. 

Why would a chain decide not to catch OOG panics and return with the appropriate result?
Why would a chain decide not to cache the state and only write state updates if msg-handler passes?

I think all chains will want to do this. If you have an example of some baseapp logic that is hardcoded but should be customizable, please share it in a reply.

I think this comes back to the trade-off between being opinionated and being flexible. Given that every chain will be doing this, I don't see why we should make every user remember to add the relevant decorators to their chain and add that unnecessary burden along with the risk that developers forget to do this.

Given that the code in `runTx` and `DeliverTx` something every SDK user should be running. I think it makes sense to be opinionated here and have it implemented directly into the `DeliverTx` implementation.

This along with my comment [here](https://github.com/cosmos/cosmos-sdk/pull/4942#discussion_r318182836) is why I think it's better to have decorators deliberately restricted to the AnteHandler, rather than have it wrap the entire Handler execution path.